src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+    }
+}
+
+interface IMatchedTokenList extends Array<MatchedToken> {}
+
+export class FoldingProvider implements vscode.FoldingRangeProvider {
+    private powershellGrammar;
+
+    public async provideFoldingRanges(
+        document: vscode.TextDocument,
+        context: vscode.FoldingContext,
+        token: vscode.CancellationToken,
+    ): Promise<vscode.FoldingRange[]> {
+
+        const foldingRanges = [];
+
+        // Convert the document text into a series of grammar tokens
+        const tokens = this.grammar().tokenizeLine(document.getText()).tokens;
+
+        // Parse the token list looking for matching tokens and return
+        // a list of IMatchedTokens.  Then filter the list and only return matches
+        // that are a valid folding range e.g. It meets a minimum line span limit
+        const foldableTokens = this.matchGrammarTokens(tokens, document)
+            .filter((item) => item.isValidRange());
+
+        // Sort the list of matched tokens, starting at the top of the document,
+        // and ensure that, in the case of multiple ranges starting the same line,
+        // that the largest range (i.e. most number of lines spanned) is sorted
+        // first.  This is needed as vscode will just ignore any duplicate folding
+        // ranges.
+        foldableTokens.sort((a: MatchedToken, b: MatchedToken) => {
+            // Initially look at the start line
+            if (a.startline > b.startline) { return 1; }
+            if (a.startline < b.startline) { return -1; }
+            // They have the same start line so now consider the end line.
+            // The biggest line range is sorted first
+            if (a.endline > b.endline) { return -1; }
+            if (a.endline < b.endline) { return 1; }
+            // They're the same
+            return 0;
+        });
+
+        // Convert the matched token list into a FoldingRange[]
+        foldableTokens.forEach((item) => { foldingRanges.push(item.toFoldingRange()); });
+
+        return foldingRanges;
+    }
+
+    // Given a start and end textmate scope name, find matching grammar tokens
+    // and pair them together.  Uses a simple FIFO stack to take into account
+    // nested regions.
+    private matchScopeElements(tokens,
+                               startScopeName: string,
+                               endScopeName: string,
+                               matchType: MatchType,
+                               document: vscode.TextDocument): IMatchedTokenList {
+        const result = [];
+        const tokenStack = [];
+
+        tokens.forEach((token) => {
+            if (token.scopes.includes(startScopeName)) {
+                tokenStack.push(token);
+            }
+            if (token.scopes.includes(endScopeName)) {
+                result.push(new MatchedToken(tokenStack.pop(), token, null, null, matchType, document));
+            }
+        });
+
+        return result.reverse();
```

Comment:
if you need to return the reversed result, maybe use `result.unshift` instead of `result.push`?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/unshift

Reply:
Fixed.